src/QirRuntime/lib/QIR/intrinsics.cpp

```diff
@@ -15,38 +15,79 @@
 #include "context.hpp"
 #include "qirTypes.hpp"
 
+// Pauli consts are {i2} in QIR, likely stored as {i8} in arrays, but we are using the standard C++ enum type based on
+// {i32} so cannot pass through the buffer and have to allocate a new one instead and copy.
+static std::vector<PauliId> ExtractPauliIds(QirArray* paulis)
+{
+    const long count = paulis->count;
+    std::vector<PauliId> pauliIds;
+    pauliIds.reserve(count);
+    for (long i = 0; i < count; i++)
+    {
+        pauliIds.push_back(static_cast<PauliId>(*paulis->GetItemPointer(i)));
+    }
+    return pauliIds;
+}
+
 extern "C"
 {
-    double quantum__qis__intAsDouble(long value)
+    void quantum__qis__exp__body(QirArray* paulis, double angle, QirArray* qubits)
+    {
+        assert(paulis->count == qubits->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
+        return Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->Exp(
+            paulis->count, reinterpret_cast<PauliId*>(pauliIds.data()), reinterpret_cast<Qubit*>(qubits->buffer),
+            angle);
+    }
+
+    void quantum__qis__exp__adj(QirArray* paulis, double angle, QirArray* qubits)
+    {
+        assert(paulis->count == qubits->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
+        return Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->Exp(
+            paulis->count, reinterpret_cast<PauliId*>(pauliIds.data()), reinterpret_cast<Qubit*>(qubits->buffer),
+            -angle);
+    }
+
+    void quantum__qis__exp__ctl(QirArray* ctls, QirArray* paulis, double angle, QirArray* qubits)
     {
-        assert(value == static_cast<int32_t>(value));
-        return static_cast<double>(value);
+        assert(paulis->count == qubits->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
+        return Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->ControlledExp(
+            ctls->count, reinterpret_cast<Qubit*>(ctls->buffer), paulis->count,
+            reinterpret_cast<PauliId*>(pauliIds.data()), reinterpret_cast<Qubit*>(qubits->buffer), angle);
     }
 
-    void quantum__qis__cnot(Qubit control, Qubit qubit)
+    void quantum__qis__exp__ctladj(QirArray* ctls, QirArray* paulis, double angle, QirArray* qubits)
     {
-        Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->ControlledX(1, &control, qubit);
+        assert(paulis->count == qubits->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
+        return Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->ControlledExp(
+            ctls->count, reinterpret_cast<Qubit*>(ctls->buffer), paulis->count,
+            reinterpret_cast<PauliId*>(pauliIds.data()), reinterpret_cast<Qubit*>(qubits->buffer), angle);
     }

```

Comment:
I don't see the difference between 
`quantum__qis__exp__ctl()` and 
`quantum__qis__exp__ctladj()`. 
May be the latter can call the former?
```c++
void quantum__qis__exp__ctladj(QirArray* ctls, QirArray* paulis, double angle, QirArray* qubits)
{
    quantum__qis__exp__ctl(ctls, paulis, angle, qubits);
}
```

Reply:
That may be a bug; the sign of the angle passed to `ControlledExp` should be inverted when calling the controlled adjoint variant.

Reply:
You are both correct. The angle should be inverted, but to make it easier to see the logic, it helps to implement the adjoint in terms of body.